ballerina/graphql_client.bal

```diff
@@ -0,0 +1,183 @@
+    # + return - An error at the failure of client initialization
+    public isolated function init(string serviceUrl, http:ClientConfiguration clientConfig = {})  returns GError? {
+        do {
+            http:Client httpEp = check new (serviceUrl, clientConfig);             
+            self.clientEp = httpEp;
+        } on fail var e {
+            return error ClientError("GraphQL Client Error", e);
+        }
+        return;
+    }
+
+    # Executes a GraphQL query and data binds the GraphQL response to a record with data and extensions 
+    # which is a subtype of GenericResponse. The errors of the GraphQL response are directed in the Ballerina error path.
+    #
+    # + query - The GraphQL query. For example `query countryByCode($code:ID!) {country(code:$code) {name}}`.
+    # + variables - The GraphQL variables. For example `{"code": "<variable_value>"}`.
+    # + headers - The GraphQL API headers to execute each query 
+    # + targetType - The payload, which is expected to be returned after data binding. For example 
+    #                `type CountryByCodeResponse record {| map<json?> __extensions?; record {| record{|string name;|}? country; |} data;`
+    # + return - The payload (if the `targetType` is configured) or a `graphql:GError` if failed to execute the query
+    remote isolated function executeWithType(string query, map<anydata>? variables = (), 
+                                             map<string|string[]>? headers = (), 
+                                             typedesc<GenericResponse|DataResponse|record{}|json> targetType = <>) 
+                                             returns targetType|GError = @java:Method {
+        'class: "io.ballerina.stdlib.graphql.client.QueryExecutor",
+        name: "executeWithType"
+    } external;
+
+    private isolated function processExecuteWithType(typedesc<GenericResponse|DataResponse|record{}|json> targetType, 
+                                                     string query, map<anydata>? variables, 
+                                                     map<string|string[]>? headers) 
+                                                     returns GenericResponse|DataResponse|record{}|json|GError {
+        http:Request request = new;
+        json graphqlPayload = getGraphqlPayload(query, variables);
+        request.setPayload(graphqlPayload);
+
+        json|http:ClientError httpResponse = self.clientEp->post("", request, headers = headers);
+
+        do {
+            if httpResponse is http:ClientError {
+                if (httpResponse is http:ApplicationResponseError) {
+                    anydata data = check httpResponse.detail().get("body").ensureType(anydata);
+                    return error ClientError("GraphQL Client Error", body = data);
+                }
+                return error ClientError("GraphQL Client Error", httpResponse);
+            } else {
```

Comment:
Since the previous type guard has returned a value, we do not need an `else` block here. It should work outside the `else` block.

Reply:
The else block was removed. The changes are amended. Thank you.

Reply:
Else block is still here, right?

Reply:
![Screenshot from 2022-03-29 17-35-22](https://user-images.githubusercontent.com/44081958/160607417-2f53dc8b-ef0f-481c-a5ed-aead2ac6fa12.png)
 Gets the following compile time error when the else block is removed

Reply:
Instead of using `check`, the previous if block should return directly since it is returning an error anyway.

Maybe this is an improvement @MaryamZi WDYT?

Reply:
I would use `return` in the preceding `if` block, since generally `check` vs` return` means two things. It help readability too IMO. Right now, I had to look at the return type of `handleGraphqlErrorResponse` to figure out that the `if` block always returns.

But I agree that narrowing should have worked with `check` also. Can y'all please create an issue for this (in ballerina-lang)?

Reply:
Created an issue in Ballerina lang https://github.com/ballerina-platform/ballerina-lang/issues/35622.
The change is applied in [47e5a38](https://github.com/ballerina-platform/module-ballerina-graphql/pull/663/commits/47e5a38537e0e31ee452d51e9943f81c773e3a95). Thank you.